datacite-delete-draft-doi-handler/src/main/java/no/unit/nva/datacite/handlers/resource/ExternalUpdatesEventHandler.java

```diff
@@ -0,0 +1,136 @@
+import no.unit.nva.doi.datacite.clients.exception.ClientException;
+import no.unit.nva.doi.datacite.restclient.models.State;
+import no.unit.nva.doi.models.Doi;
+import no.unit.nva.events.models.AwsEventBridgeDetail;
+import no.unit.nva.events.models.AwsEventBridgeEvent;
+import no.unit.nva.events.models.EventReference;
+import no.unit.nva.s3.S3Driver;
+import nva.commons.core.Environment;
+import nva.commons.core.JacocoGenerated;
+import nva.commons.core.attempt.Failure;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.s3.S3Client;
+
+public class ExternalUpdatesEventHandler implements RequestHandler<SQSEvent, Void> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ExternalUpdatesEventHandler.class);
+    private static final TypeReference<AwsEventBridgeEvent<AwsEventBridgeDetail<EventReference>>>
+        SQS_VALUE_TYPE_REF = new TypeReference<>() {
+    };
+    private static final String RESOURCE_DELETED_TOPIC = "PublicationService.Resource.Deleted";
+    private static final String EVENTS_BUCKET_NAME_ENV = "EVENTS_BUCKET_NAME";
+    private static final String REMOVE_ACTION = "REMOVE";
+
+    private final S3Driver s3Driver;
+    private final DoiClient doiClient;
+
+    @JacocoGenerated
+    public ExternalUpdatesEventHandler() {
+        this(
+            new Environment(),
+            S3Driver.defaultS3Client().build(),
+            defaultDoiClient());
+    }
+
+    protected ExternalUpdatesEventHandler(
+        Environment environment, S3Client s3Client, DoiClient doiClient) {
+        this.s3Driver = new S3Driver(s3Client, environment.readEnv(EVENTS_BUCKET_NAME_ENV));
+        this.doiClient = doiClient;
+    }
+
+    @Override
+    public Void handleRequest(SQSEvent sqsEvent, Context context) {
+        Optional.ofNullable(sqsEvent.getRecords()).stream()
+            .flatMap(List::stream)
+            .map(ExternalUpdatesEventHandler::parseEventReference)
+            .filter(Objects::nonNull)
+            .forEach(this::processPayload);
+        return null;
+    }
+
+    private void processPayload(EventReference eventReference) {
+        verifyTopic(eventReference);
+        var updateEvent = getEventBodyFromS3(eventReference);
+        verifyAction(updateEvent);
+        deleteDraftDoiIfPresent(updateEvent);
+    }
+
+    private void deleteDraftDoiIfPresent(UpdateEvent updateEvent) {
+        var customerId = updateEvent.oldData().publisher().id();
+        if (nonNull(updateEvent.oldData().doi())) {
+            var doi = Doi.fromUri(updateEvent.oldData().doi());
+            try {
+                var doiState = doiClient.getDoi(customerId, doi);
+                if (State.DRAFT.equals(doiState.getState())) {
+                    doiClient.deleteDraftDoi(customerId, doi);
+                    logger.info("Deleted draft DOI {} as resource {} was deleted.", doi,
+                                updateEvent.oldData().identifier());
```

Comment:
This relates to Kir's comment…probably extract the code and give it a suitable name for extensibility if what Kir is asking tracks with what you plan to do later.

Reply:
Also, take the ClientException throwing code into a separate block, so we don't get update errors thrown as client exceptions.